src/main/java/org/veupathdb/service/eda/ss/model/reducer/SubsettingJoinNode.java

```diff
@@ -1,24 +1,60 @@
 package org.veupathdb.service.eda.ss.model.reducer;
 
+import org.veupathdb.service.eda.ss.model.Entity;
+import org.veupathdb.service.eda.ss.model.Study;
+import org.veupathdb.service.eda.ss.model.reducer.ancestor.AncestorMapperFactory;
+import org.veupathdb.service.eda.ss.model.variable.VariableValueIdPair;
+
 import java.util.ArrayList;
 import java.util.Iterator;
 import java.util.List;
+import java.util.stream.Collectors;
 
+/**
+ * Node representing an entity in the map-reduce entity tree. The structure of the tree is a function of the study's
+ * entity hierarchy and the output variable. Opens value streams for all variables on which we are filtering for this
+ * entity and intersects all of the ID indexes that are not filtered. Recursively reduce each child node in this manner
+ * and map the child ID Indexes to this element's ID indexes.
+ */
 public class SubsettingJoinNode {
+  private final List<Iterator<Long>> filters; // ID indexes of this Entity's "type" for filtered data streams.
+  private final List<SubsettingJoinNode> children;
+  private final Entity entity;
+  private final Study study;
+  private final AncestorMapperFactory ancestorMapperFactory;
 
-  private final List<FilteredValueFile<?, Long>> filters;
-
-  public SubsettingJoinNode(List<FilteredValueFile<?, Long>> filters) {
-    this.filters = filters;
+  public SubsettingJoinNode(List<Iterator<Long>> filteredStreams,
+                            List<SubsettingJoinNode> children,
+                            Entity entity,
+                            Study study,
+                            AncestorMapperFactory ancestorMapperFactory) {
+    this.filters = filteredStreams;
+    this.children = children;
+    this.entity = entity;
+    this.study = study;
+    this.ancestorMapperFactory = ancestorMapperFactory;
   }
 
   public Iterator<Long> reduce() {
-    // TODO include children.
-    final List<Iterator<Long>> idStreams = new ArrayList<>();
-    for (FilteredValueFile<?, Long> file: filters) {
-      idStreams.add(file);
+    List<Iterator<Long>> childStreams = children.stream()
```

Comment:
this tricky recursive logic could benefit from a comment.

Reply:
Added a comment